### ISSUE TITLE
Extending / modifying the character set

### ISSUE THREAD

**Issue by jublo** [source-foundry/Hack#307]
Currently, the goal is to support web font subsets of ASCII, Latin-1, and Latin extended A – a decision by @burodepeper and @chrissimpkins.

Would be worth having a more detailed discussion about this and reviewing open IR for other characters to see what is important to the team (e.g. math glyphs?) and what is not?

*Note:* This may require modification of design document based upon the conversation.

**Comment by burodepeper** [comment]
The core character set is as you've described, with the addition of some currency and punctuation glyphs. My estimate is that including those characters (or a similar set like MES-1, I believe) covers about 99% of use cases. This decision was largely based on my personal assumptions.

I'd like this issue to focus more on a better definition of what glyphs should be considered to be the core of Hack, and what (groups of) glyphs are good extensions that can be considered as options when building a custom version of your (web)font. The rationale behind that is that the core team should prioritize the core set of glyphs, with the possibility of extensions (including alternate glyphs) being maintained by others.

**Comment by jublo** [comment]
Reference to certain character sets: https://codepoints.net/basic_multilingual_plane

**Comment by chrissimpkins** [comment]
👍 much needed discussion!  I take full responsibility for the mess that is the differences in extended glyph support in the four variants and we should take a good hard look at what is:

- important to support source code (currently felt to be ASCII)
- important to support source code associated areas (language support for comments, string literals, etc) - @jublo has done work on our character set support in the past and it would be helpful to review where we are with these tools so that we can have a starting point for the discussion
- important to strike proper balance of font size vs. set support in web font/web font subsets.  A separate but related issue where file size is something that should be much more highly scrutinized

Let's get the conversation going!

**Comment by chrissimpkins** [comment]
And as an added note to @burodepeper comments above, we have to be able to draw the line somewhere on set support.  Largely depends upon hands available for the work, what we intend to achieve with Hack (i.e. breadth of areas that we want this monospaced typeface to support), and whether some issues belong in derivatives / optional compiles and the like.

**Comment by jublo** [comment]
Once this discussion led to results, we can analyse which character sets have gaps and/or move glyphs to the Hack buffet (a place where users can pick what they want).

**Comment by burodepeper** [comment]
It's currently advertised that Hack includes 1561 glyphs, yet only a little over 300 ended up in the webfont.

**Comment by chrissimpkins** [comment]
We should also differentiate what is important to have in the sets and require further optimization beyond the design as submitted in a commit (i.e. relative shape/spacing - including hinting - considerations vs other glyphs in set) from those that are felt to be useful to have but should not be priorities for this type of work. Currently, everything outside of ASCII falls into the latter category.  It's good that it is there because people use it for non-source code text glyph coverage, but not what we are attempting to optimize here.

**Comment by chrissimpkins** [comment]
> It's currently advertised that Hack includes 1561 glyphs, yet only a little over 300 ended up in the webfont.

You are looking at the subset web font I believe.  There is a full set as well

**Comment by burodepeper** [comment]
> You are looking at the subset web font I believe. There is a full set as well

True, but in my experience webfonts rarely use a full set.

**Comment by chrissimpkins** [comment]
> True, but in my experience webfonts rarely use a full set.

Misunderstood you.  Thought you meant that there was not a set that included all glyphs.

**Comment by chrissimpkins** [comment]
See my comment in https://github.com/source-foundry/Hack/issues/308#issuecomment-329474773 re: extended character set approach.  Extending the sets with an external repository that is intended to be used to pull in "extensions" may prove problematic and make things more difficult for our redistribution teams.  There are lots of considerations downstream that this decision will potentially impact.  I suggest that we use the conversation in this thread to identify how to approach extended character sets within the Hack source and whether further extended sets would be more appropriate as forks of Hack that are maintained separately under the same build tooling rather than creating a new problem that requires new build tooling to address.  Forks would be simple.  Fork, add new glyphs, pull any upstream changes to incorporate them there.  This would work with current build approach.  Extensions to the source would not.

I think that the external glyph repository is better suited to replacements of existing glyphs within the sets with alternate designs. (IMO)

**Comment by chrissimpkins** [comment]
And to provide a correction to the OP, we did not make a decision to limit Hack to those sets.  The decision was to limit the web font subset builds to those sets.

**Comment by jublo** [comment]
Based on @texhex’s input on #308, we need to decide and draw a line between glyphs supported and maintained by ourselves, and ones less tested/supported/biased by the community (suggested as `contrib` directory). What should become core and what not?

**Comment by jublo** [comment]
I propose the following – tell your opinion:

* Support ASCII, Latin, Latin-Extended A, MES 1, MES 2 in Hack Core.
* Anything else goes to `contrib`.
* This is not to exclude, but to graduate popular contributions with their own maintainers.
   * For example, there could be a popular “overlay” `powerline` with someone maintaining it in a fork and sending PRs.

**Comment by burodepeper** [comment]
> Support ASCII, Latin, Latin-Extended A, MES 1, MES 2 in Hack Core

Can you summarize what this includes in general? Or perhaps more specifically, which glyphs currently in the full set would it exclude?

**Comment by chrissimpkins** [comment]
> Can you summarize what this includes in general? Or perhaps more specifically, which glyphs currently in the full set would it exclude?

Maybe link those sets as images from your character set coverage tool?  I agree that it would be helpful to see what we are talking about including as well as excluding based upon what currently exists in the sets.  It would also be helpful to see where we currently are with coverage relative to other commonly used monospaced typefaces out there (I think that this is implemented in your tool also and seem to recall you showing these data a couple of years ago?).   With this information we can decide whether there are areas to expand (e.g. to achieve complete coverage of a set felt to be important in a monospaced face), whether there are areas to cut (e.g. there is definitely some detritus in each of the sets that will never be rendered), and what is perhaps an coverage attribute for Hack relative to other monospaced typefaces if we take into consideration non-ASCII uses for the fonts.  While not a primary **source code design optimization priority** for us it does not mean that these sets are unimportant to users out there, for source code associated use cases such as string literals and comments or otherwise, and that we should cast them away.  Given that there is absolutely no maintenance requirement for things that already exist in the sets their removal creates more work than simply leaving them be with little downside that I can see in the desktop setting if there is a use case that we can identify.  Web fonts are a different issue and the current conversation does not appear to be focused on web fonts and file size saving considerations but rather what has come to be called "Core" which I am reading as our decision about the overall default font set available through the main repository.  There are going to be sets that likely fall outside of the definition above (including sets like the box drawing glyphs) that are important to some users.  I suggest that we maintain and expand (as a future down the road issue) the mathematics/stats glyphs.  These don't need the attention to source code tuning that we are giving to the ASCII sets but they are valuable in a monospaced typeface.  We will know once we examine everything that is there.  Web fonts will need more stringent criteria applied to balance the size savings issue and I think that these should be different criteria than those that are applied to the desktop fonts.

> Do we want alternates (to existing glyphs in Core or contrib) in a separate repo, or do we want them in this central repo?

This is largely going to depend upon the build tool that is used to accomplish the alternates approach.  I would keep the conversation about alternate glyphs in the #308 thread as this IR is about core character set coverage, not alternate glyphs and we do not want to confuse the issues.  The association between the two conversations will be that there cannot be an alternate for something that does not exist in the core sets, this would by definition be an extension instead and may require a different build approach.

> This is not to exclude, but to graduate popular contributions with their own maintainers

I am not certain what you mean here.  Are you suggesting that the contrib area of the repository is one in which we are suggesting interested parties consider development and maintenance of sets of glyphs that we take no interest in?  I'm not sure (1) why we would want to maintain such glyphs in the main repository ourselves once such a decision is made; (2) that if we did it warrants source code isolation in a separate directory structure?  There is absolutely no maintenance involved in keeping them in the main body of source code.  To maintain a separate set of glyph designs that will open in a font editor for work, we would need to create a new valid UFO spec directory structure for these glyphs that are isolated from the main source.  This generates more work for something that we are defining as unimportant.  I suggest that we maintain a single body of source in the repository and not begin to fragment it.
